OpenSim/Simulation/SimbodyEngine/Joint.cpp

```diff
@@ -655,6 +655,84 @@ void Joint::updateFromXMLNode(SimTK::Xml::Element& aNode, int versionNumber)
             }
         }
 
+        // Version 30514 removed the user-facing "reverse" property from Joint.
+        // The parent and child frames are swapped if a "reverse" element is
+        // found and its value is "true".
+        if (documentVersion < 30514) {
+            auto reverseElt = aNode.element_begin("reverse");
+
+            if (reverseElt != aNode.element_end()) {
+                bool swapFrames = false;
+                reverseElt->getValue().tryConvertToBool(swapFrames);
+
+                if (swapFrames) {
+                    std::string oldParentFrameName = "";
+                    std::string oldChildFrameName  = "";
+
+                    // Find names of parent and child frames. If more than one
+                    // "parent_frame" or "child_frame" element exists, keep the
+                    // first one. The "parent_frame" and "child_frame" elements
+                    // may be listed in either order.
+                    SimTK::Xml::element_iterator connectorsNode =
+                        aNode.element_begin("connectors");
+                    SimTK::Xml::element_iterator connectorElt = connectorsNode->
+                        element_begin("Connector_PhysicalFrame_");
```

Comment:
Yeah, that was a source of confusion. I think the current design might be better than duplicating the code. Better still might be splitting up the code so that the model file is completely updated one version at a time. We might also then permit updates from an old version to a version that isn't current, which could help test backwards compatibility.